axiom/nr_txn.c

```diff
@@ -3293,6 +3297,90 @@ bool nr_txn_log_decorating_enabled(nrtxn_t* txn) {
   return false;
 }
 
+#define ENSURE_LOG_LEVEL_NAME(level_name) \
+  (NULL == level_name ? "UNKNOWN" : level_name)
+
+static void log_event_set_linking_metadata(nr_log_event_t* e,
+                                           nrtxn_t* txn,
+                                           nrapp_t* app) {
+  char* trace_id = NULL;
+  char* span_id = NULL;
+
+  if (NULL == e) {
+    return;
+  }
+
+  if (nrlikely(txn)) {
+    trace_id = nr_txn_get_current_trace_id(txn);
+    nr_log_event_set_trace_id(e, trace_id);
+    nr_free(trace_id);
+
+    span_id = nr_txn_get_current_span_id(txn);
+    nr_log_event_set_span_id(e, span_id);
+    nr_free(span_id);
+
+    nr_log_event_set_entity_name(e, txn->primary_app_name);
+  }
+
+  if (nrlikely(app)) {
+    nr_log_event_set_hostname(e, nr_app_get_host_name(app));
+    nr_log_event_set_guid(e, nr_app_get_entity_guid(app));
+  }
+}
+
+static nr_log_event_t* log_event_create(const char* log_level_name,
+                                        const char* log_message,
+                                        nrtime_t timestamp,
+                                        nrtxn_t* txn,
+                                        nrapp_t* app) {
+  nr_log_event_t* e = nr_log_event_create();
+  nr_log_event_set_log_level(e, ENSURE_LOG_LEVEL_NAME(log_level_name));
+  nr_log_event_set_message(e, log_message);
+  nr_log_event_set_timestamp(e, timestamp);
+
+  log_event_set_linking_metadata(e, txn, app);
+
+  return e;
+}
+
+static void nr_txn_add_log_event(nrtxn_t* txn,
+                                 const char* log_level_name,
+                                 const char* log_message,
+                                 nrtime_t timestamp,
+                                 nrapp_t* app) {
+  nr_random_t* rnd = NULL;
+  nr_log_event_t* e = NULL;
+  bool event_dropped = false;
+
+  if (!nr_txn_log_forwarding_enabled(txn)) {
```

Comment:
This causes an interesting problem - since nr_txn_log_forwarding_enabled() will return false if the max log samples is set to 0 then we have a case where the "dropped" metric isn't updated (which happens at the end of this function).
If the forwarding flag is enabled but max samples set to 0 it seems like we are dropping log lines and it should be reflected.

Reply:
@mfulb Nice find! I agree. I think the solution to this would be to remove this check:
```
  if (0 == txn->options.log_events_max_samples_stored) {
    return false;
  }
```
from `nr_txn_log_forwarding_enabled` function and allow `txn->log_events` pool to have the size of 0. This way `nr_analytics_events_add_event` would correctly return information that an event is dropped. Thoughts?

Reply:
I really like having the pool size of 0 return false from `nr_txn_log_forwarding_enabled` as it will guarantee nothing gets added even if the pool was created accidentally with a non-zero size.
But it does complicate the implementation here.

Reply:
@mfulb Look at [axiom/nr_txn.c#L495](https://github.com/misiekn/newrelic-php-agent/blob/e7f4dcb489b4b7659732b83994157838aff1a876/axiom/nr_txn.c#L495) This is how it's done for other types of events - the `<event_type>_enabled` flag is set based on agent setting and app limits. This seems to be in line with what `nr_txn_log_forwarding_enabled` is currently doing (with `if (0 == txn->options.log_events_max_samples_stored) return false;` check).

Reply:
Support for log event's pool of size 0 is added with 2f54982c76d39280b5c946062eb050dd25e12f8f, 486ce76a699939e65c40d50fdcc3d9affa6063ec, 4e17dc09758b5455fcf6a8e8dc37b48fea497886. 09c96740de6b1e674508510b55e85178215a4c58 removes dependency of log forwarding feature on the log event's pool size. This way when event will be added to a pool of size 0 it will be marked as **seen** and **dropped**.

Note: I decided not to 'short-circuit' `nr_txn_add_log_event` when log event pool has the size 0. This way the log event pool has the **seen** counter updated too.